File: src/main/java/com/codingforcookies/betterrecords/block/ModBlocks.java
package com.codingforcookies.betterrecords.block;

import com.codingforcookies.betterrecords.block.tile.*;
import net.minecraft.block.Block;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.fml.common.registry.GameRegistry;

public final class ModBlocks {

    public static Block blockRecordEtcher;
    public static Block blockRecordPlayer;
    public static Block blockFrequencyTuner;
    public static Block blockRadio;

    public static Block blockSMSpeaker;
    public static Block blockMDSpeaker;
    public static Block blockLGSpeaker;

    public static Block blockStrobeLight;
    public static Block blockLazer;
    public static Block blockLazerCluster;

    public static void init() {
        blockRecordEtcher = new BlockRecordEtcher("recordetcher");
        blockRecordPlayer = new BlockRecordPlayer("recordplayer");
        blockFrequencyTuner = new BlockFrequencyTuner("frequencytuner");
        blockRadio = new BlockRadio("radio");

        blockSMSpeaker = new BlockSpeaker("speaker.sm", 0);
        blockMDSpeaker = new BlockSpeaker("speaker.md", 1);
        blockLGSpeaker = new BlockSpeaker("speaker.lg", 2);

        blockStrobeLight = new BlockStrobeLight("strobelight");
        blockLazer = new BlockLazer("lazer");
        blockLazerCluster = new BlockLazerCluster("lazercluster");

        initTileEntities();
    }

    private static void initTileEntities() {
        registerTile(TileRecordEtcher.class, "recordetcher");
        registerTile(TileRecordPlayer.class, "recordplayer");
        registerTile(TileFrequencyTuner.class, "frequencytuner");
        registerTile(TileRadio.class, "radio");
        registerTile(TileSpeaker.class, "speaker");
        registerTile(TileStrobeLight.class, "strobelight");
        registerTile(TileLazer.class, "lazer");
        registerTile(TileLazerCluster.class, "lazercluster");
    }

    private static void registerTile(Class<? extends TileEntity> class_, String key) {
        GameRegistry.registerTileEntity(class_, "betterrecords:" + key);
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/block/ModBlocks.kt
package com.codingforcookies.betterrecords.block

import com.codingforcookies.betterrecords.ID
import com.codingforcookies.betterrecords.block.itemblock.ItemBlockSpeaker
import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraft.item.ItemBlock
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

@Mod.EventBusSubscriber(modid = ID)
object ModBlocks {

    val blockRecordEtcher = BlockRecordEtcher("recordetcher")
    val blockRecordPlayer = BlockRecordPlayer("recordplayer")
    val blockFrequencyTuner = BlockFrequencyTuner("frequencytuner")
    val blockRadio = BlockRadio("radio")
    val blockSpeaker = BlockSpeaker("speaker")
    val blockStrobeLight = BlockStrobeLight("strobelight")
    val blockLaser = BlockLaser("laser")
    val blockLaserCluster = BlockLaserCluster("lasercluster")

    // TODO: Use reflection to register declared fields

    @JvmStatic
    @SubscribeEvent
    fun registerBlocks(event: RegistryEvent.Register<Block>) {
        event.registry.registerAll(
                blockRecordEtcher,
                blockRecordPlayer,
                blockFrequencyTuner,
                blockRadio,
                blockSpeaker,
                blockStrobeLight,
                blockLaser,
                blockLaserCluster
        )
    }

    @JvmStatic
    @SubscribeEvent
    fun registerItems(event: RegistryEvent.Register<Item>) {
        arrayOf<ModBlock>(
                blockRecordEtcher,
                blockRecordPlayer,
                blockFrequencyTuner,
                blockRadio,
                blockStrobeLight,
                blockLaser,
                blockLaserCluster
        ).map {
            ItemBlock(it).setRegistryName(it.registryName)
        }.forEach(event.registry::register)

        // Register the speaker specially since it has variants
        event.registry.register(ItemBlockSpeaker(blockSpeaker).setRegistryName(blockSpeaker.registryName))

        Block.REGISTRY
                .filterIsInstance<ModBlock>()
                .forEach {
                    if (it is TileEntityProvider<*>) {
                        it.registerTileEntity(it)
                    }
                }
    }

    @JvmStatic
    @SubscribeEvent
    fun registerModels(event: ModelRegistryEvent) {
        Block.REGISTRY
                .filterIsInstance<ModBlock>()
                .forEach {
                    it.setStateMapper()

                    if (it is ItemModelProvider) {
                        it.registerItemModel(it)
                    }

                    if (it is TESRProvider<*>) {
                        it.bindTESR()
                        it.registerTESRItemStacks(it)
                    }
                }
    }
}
